### PR TITLE
9.30 Project

### PR DESCRIPTION
**Pull request by gsy19971111**

### CODE REVIEW

File: jointContribution/AI_Climate_disease/XGBoost_imputer.ipynb
Comment (wangguan1995) [inline jointContribution/AI_Climate_disease/XGBoost_imputer.ipynb:1]: 多余文件需要清理一下

File: jointContribution/AI_Climate_disease/结项报告.md
Comment (wangguan1995) [inline jointContribution/AI_Climate_disease/结项报告.md:1]: 需要按照代码规范进行修改：

1. 整理相关requirements
2. 修改ipynb为py文件，进行代码的整理
3. 使用pre-commit进行代码清理
4. 补充文档，暂时放在当前文件夹
image

https://paddlescience-docs.readthedocs.io/zh-cn/latest/zh/development/#4

**Review by wangguan1995 — CHANGES_REQUESTED** [review]
tofix

File: jointContribution/AI_Disease_Climate/era5_land_dataset.py
Comment (HydrogenSulfate) [inline jointContribution/AI_Disease_Climate/era5_land_dataset.py:1]: 这个文件还麻烦放到examples/下面，然后使用这种方法注册到ppsci.data里，可以给XXXConstraint初始化时使用

```py
## examples/era5_land_dataset.py‎

from ppsci.data import register_to_dataset

class ToyTwoModalDataset(Dataset):
    pass


register_to_dataset(ToyTwoModalDataset) # here

```

File: jointContribution/AI_Disease_Climate/TabM.py [hunk 1]
```diff
@@ -0,0 +1,223 @@
+import math
+from typing import Literal
+from typing import Optional
+
+import paddle
+import paddle.nn as nn
+
+
+# =========================
```
Comment (wangguan1995) [inline jointContribution/AI_Disease_Climate/TabM.py]: 注释清理

File: jointContribution/AI_Disease_Climate/TabM.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+    with paddle.no_grad():
+        p = paddle.full(w.shape, 0.5, dtype="float32")
+        s = paddle.bernoulli(p) * 2.0 - 1.0
+        s = paddle.cast(s, w.dtype)
+        w.set_value(s)
+    return w
+
+
+class NLinear(nn.Layer):
+    """PackedEnsemble: K 份 Linear 打包 → 输入 (B,K,D), 权重 (K, I, O)"""
```
Comment (wangguan1995) [inline jointContribution/AI_Disease_Climate/TabM.py]: 中文->英文注释

File: jointContribution/AI_Disease_Climate/TabM.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+    y_e = ((x * r_e) @ W) * s_e + b_e
+    x: (B,K,I) → y: (B,K,O)
+    """
+
+    def __init__(
+        self, in_f: int, out_f: int, k: int, scale_init="ones", bias: bool = True
+    ):
+        super().__init__()
+        self.k, self.in_f, self.out_f = k, in_f, out_f
+        self.weight = self.create_parameter(shape=[in_f, out_f])  # 共享权重
```
Comment (wangguan1995) [inline jointContribution/AI_Disease_Climate/TabM.py]: 中文注释

File: jointContribution/AI_Disease_Climate/TabM.py [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+
+    def forward(self, x_num: paddle.Tensor):  # x_num: (B, D)
+        if self.k > 1:
+            x = x_num.unsqueeze(1).tile([1, self.k, 1])  # (B,K,D)
+        else:
+            x = x_num.unsqueeze(1)  # (B,1,D)
+        if self.min_adapter is not None:
+            x = self.min_adapter(x)
+        feats = self.backbone(x)  # (B,K,H)
+        return feats.mean(axis=1) if self.reduce else feats  # (B,H) 或 (B,K,H)
```
Comment (wangguan1995) [inline jointContribution/AI_Disease_Climate/TabM.py]: 同上

File: ppsci/loss/bce.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+                loss = loss.mean()
+
+            # final global/per-key weighting like in L2Loss
+
+            losses[key] = loss
+        print(losses)
+        return losses
+
+
+# ------------------------------------------------------------
```
Comment (wangguan1995) [inline ppsci/loss/bce.py]: 此处注释删掉

File: jointContribution/AI_Disease_Climate/model.py [hunk 6]
```diff
@@ -0,0 +1,750 @@
+# -*- coding: utf-8 -*-
+import math
+
+import numpy as np
+import paddle
+import paddle.nn as nn
+import paddle.nn.functional as F
+from paddle.io import DataLoader
+
+# 选中第 0 张 GPU；如有多卡改成 'gpu:1' 等
```
Comment (wangguan1995) [inline jointContribution/AI_Disease_Climate/model.py]: 删除多余中文注释

File: jointContribution/AI_Disease_Climate/model.py [hunk 7]
```diff
@@ -0,0 +1,750 @@
...
+        pe[:, 0::2] = np.sin(position * div_term)
+        pe[:, 1::2] = np.cos(position * div_term)
+        self.register_buffer("pe", paddle.to_tensor(pe), persistable=False)
+
+    def forward(self, x):  # (B,T,D)
+        T = x.shape[1]
+        return x + self.pe[:T, :]
+
+
+# ====================== TabM（占位，可换你的实现） ======================
```
Comment (wangguan1995) [inline jointContribution/AI_Disease_Climate/model.py]: 同上

File: jointContribution/AI_Disease_Climate/model.py [hunk 8]
```diff
@@ -0,0 +1,750 @@
...
+            nn.Linear(d_hidden, d_hidden),
+            nn.ReLU(),
+        )
+        self.d_hidden = d_hidden
+
+    def forward(self, x_num: paddle.Tensor):
+        return self.net(x_num)
+
+
+# ====================== 3D ResNet-18 体数据特征抽取 ======================
```
Comment (wangguan1995) [inline jointContribution/AI_Disease_Climate/model.py]: 同上

File: jointContribution/AI_Disease_Climate/model.py [hunk 9]
```diff
@@ -0,0 +1,750 @@
...
+        self.drop = nn.Dropout(dropout)
+        self.out_dim = self.backbone.out_dim  # 512
+
+    def forward(self, x):  # (B, C, D, H, W)
+        x = self.backbone(x)
+        x = self.drop(x)
+        return x
+
+
+# ====================== MoE（Top-k；gather_nd 选择专家） ======================
```
Comment (wangguan1995) [inline jointContribution/AI_Disease_Climate/model.py]: 同上

File: jointContribution/AI_Disease_Climate/model.py [hunk 10]
```diff
@@ -0,0 +1,750 @@
...
+    def __init__(self, d_model=512, cfg: MoEConfig = None):
+        super().__init__()
+        self.moe = MoE(d_model, cfg or MoEConfig())
+
+    def forward(self, tok, domain_id=None):
+        y, aux = self.moe(tok.unsqueeze(1), domain_id=domain_id)
+        return y.squeeze(1), aux
+
+
+# ====================== Self-Attention Transformer（可 MoE） ======================
```
Comment (wangguan1995) [inline jointContribution/AI_Disease_Climate/model.py]: 同上

File: jointContribution/AI_Disease_Climate/model.py [hunk 11]
```diff
@@ -0,0 +1,750 @@
...
+    def forward(self, x, domain_id=None):
+        x = self.pos(x)
+        aux_total = 0.0
+        for layer in self.layers:
+            x, aux = layer(x, domain_id=domain_id)
+            aux_total += aux
+        return x, aux_total
+
+
+# ====================== AFNO(1D) + MoE FFN ======================
```
Comment (wangguan1995) [inline jointContribution/AI_Disease_Climate/model.py]: 同上

File: jointContribution/AI_Disease_Climate/model.py [hunk 12]
```diff
@@ -0,0 +1,750 @@
...
+class AFNO1DLayer(nn.Layer):
+    """
+    自适应傅里叶算子（时间 1D 版）：
+    - 对 (B,T,D) 沿 T 做 rFFT → (B,D,F)
```
Comment (wangguan1995) [inline jointContribution/AI_Disease_Climate/model.py]: 同上

File: ppsci/data/dataset/era5_land_dataset.py [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+        Wv = rng.normal(size=(C, 4))
+        Wt = rng.normal(size=(424, 4))
+        logits = vid_avg @ Wv + self.vec @ Wt + rng.normal(scale=0.5, size=(n, 4))
+        probs = 1.0 / (1.0 + np.exp(-logits))
+        self.y = (probs > 0.5).astype("float32")
+
+    def __getitem__(self, idx: int):
+        ###返回三个字典
+        return {"video": self.video[idx], "vec": self.vec[idx]}, {"y": self.y[idx]}, {}
+        # return self.video[idx], self.vec[idx], self.y[idx]
```
Comment (wangguan1995) [inline ppsci/data/dataset/era5_land_dataset.py]: 删除无用注释

File: ppsci/data/dataset/era5_land_dataset.py [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+
+    def __len__(self):
+        return self.n
+
+    ###在这个基础上加constraint，构建已有约束，Loss Focal BCE，valid_dataloader_cfg,加input和label
```
Comment (wangguan1995) [inline ppsci/data/dataset/era5_land_dataset.py]: 同上

**Review by wangguan1995 — CHANGES_REQUESTED** [review]
to fix

File: ppsci/loss/bce.py [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+            raise ValueError(
+                f"reduction should be 'mean' or 'sum', but got {reduction}"
+            )
+        super().__init__(reduction, weight)
+
+    def forward(
+        self, output_dict, label_dict, weight_dict=None
+    ) -> Dict[str, "paddle.Tensor"]:
+        losses: Dict[str, paddle.Tensor] = {}
+        # print(label_dict)
```
Comment (wangguan1995) [inline ppsci/loss/bce.py]: 清理一下print

File: ppsci/loss/bce.py [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+            # reduction over batch
+            if self.reduction == "sum":
+                loss = loss.sum()
+            else:  # "mean"
+                loss = loss.mean()
+
+            # final global/per-key weighting like in L2Loss
+
+            losses[key] = loss
+        print(losses)
```
Comment (wangguan1995) [inline ppsci/loss/bce.py]: 清理一下print

**Review by wangguan1995 — CHANGES_REQUESTED** [review]
to fix

**Review by mattheliu — COMMENTED** [review]
尽快用pre-commit过一下代码规范CI

**Review by HydrogenSulfate — APPROVED** [review]
